### PR TITLE
Add linear eltwise fusion for cpu in inference mode

### PR DESCRIPTION
**Pull request by chunyuan-w**
An FX transformation is added to fuse linear with eltwise OPs in torchinductor for CPU in inference mode.

The fusion OP is implemented in https://github.com/pytorch/pytorch/pull/85622 and will be treated as an extern kernel call in torchinductor.

The fusion of linear with the below OPs is supported:

- relu
- sigmoid
- tanh
- hardswish
- leaky_relu
- hardtanh
- gelu

### CODE REVIEW

File: torchinductor/ir.py [hunk 1]
```diff
@@ -2840,6 +2840,49 @@ def __init__(self, layout, input, index):
     def should_allocate(self):
         return False
 
+class LinearReLU(ExternKernelAlloc):
```
Comment (EikanWang) [inline torchinductor/ir.py:2843]: `LinearReLU` is specific for `Linear` + `ReLU`. Should this class be `LinearEltwise`?

File: torchinductor/overrides.py [hunk 2]
```diff
@@ -163,3 +230,15 @@ def rand_like(x, **kwargs):
 
 
 replacements = {torch.nn.functional.dropout: lowmem_dropout, torch.rand_like: rand_like}
+
+computation_op = nn.Linear
```
Comment (EikanWang) [inline torchinductor/overrides.py:234]: Currently, the computation operator only covers `Linear`. In the future, we also need to support `Conv` and `Matmul`. Should the `computation_op` be a list?

**Comment by EikanWang** [comment]
@jansel , @desertfire , this PR demonstrates how to integrate oneDNN post-op fusions in TorhcInductor. Please feel free to comment on the design.

File: test/test_linear_fusion_cpu.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        nn.Tanh(),
+        nn.Hardswish(),
+        nn.LeakyReLU(0.1, inplace=False),
+        nn.Hardtanh(min_val=-0.5, max_val=4, inplace=False),
+        nn.GELU(approximate="none"),
+        nn.GELU(approximate="tanh"),
+    ]
+    return eltwise_list
+
+class TestFuseFx(QuantizationTestCase):
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply test/test_linear_fusion_cpu.py:25]: I want leverage this [checkGraphModuleNodes](https://github.com/pytorch/pytorch/blob/ea81138bd6b688554ad307e370bebeffd264f1b7/torch/testing/_internal/common_quantization.py#L583) function to check whether the fusion is successful or not and this function is defined in the [QuantizationTestCase](https://github.com/pytorch/pytorch/blob/ea81138bd6b688554ad307e370bebeffd264f1b7/torch/testing/_internal/common_quantization.py#L332) class.

File: torchinductor/ir.py (hunk 1 above)
Comment (chunyuan-w) [reply torchinductor/ir.py]: Fixed the name.

File: test/test_linear_fusion_cpu.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jgong5) [reply test/test_linear_fusion_cpu.py:25]: How about we add note to explain why we have to inherit `QuantizationTestCase` here? We may consider to move `checkGraphModuleNodes` util function to a common place later to avoid confusion.

File: torchinductor/ir.py [hunk 4]
```diff
@@ -2840,6 +2840,49 @@ def __init__(self, layout, input, index):
     def should_allocate(self):
         return False
 
+class LinearEltwise(ExternKernelAlloc):
+    kernel = "torch.ops.mkldnn_prepacked.linear_eltwise"
```
Comment (jgong5) [inline torchinductor/ir.py:2844]: About the naming: Is the post-op unary only? Also, to be consistent with existing IR names, would "pointwise" be better than "eltwise"?

File: torchinductor/overrides.py [hunk 5]
```diff
@@ -36,6 +39,70 @@ def replace_fx(gm: torch.fx.GraphModule):
...
+                    eltwise,
+                    op_name,
+                    op_info,
+                    linear.in_features,
+                    linear.out_features,
+                    linear.bias is not None,
+                    linear.weight.device,
+                    linear.weight.dtype)
+
+def fuse_fx(gm: torch.fx.GraphModule, example_inputs):
```
Comment (jgong5) [inline torchinductor/overrides.py:83]: Not sure if it is proper to add backend-specific FX passes here. Would it better to add a FX pass registration mechanism here (e.g., a registration decorator similar to how lowerings are registered) and move the backend-specific passes to a separate backend file?

File: test/test_linear_fusion_cpu.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply test/test_linear_fusion_cpu.py:28]: Comments are added.

File: torchinductor/overrides.py (hunk 2 above)
Comment (chunyuan-w) [reply torchinductor/overrides.py]: Refactored the code to be more extensible to add support for Conv and Matmul in the future.

**Comment by soumith** [comment]
ideally, ideep could have a python API + package, or directly be a C++ dependency in Inductor.
I don't like the separation of bindings, where for every fusion you want to implement, you will send a PR to pytorch/pytorch and then hook up that pattern-match into inductor.
Inductor's strength isn't in pattern-matching to fast kernels, but is to generate code dynamically.
For example, it's a purely artificial restriction that the epilogue only allows the non-linearities specified above.

Also, the blackbox nature here is a bit uncomfortable, we would ideally have inductor over time have C++ generated code for CPU. A lowered IR can be passed to a powerful "CPU compiler" for example -- like LLVM. It's not ideal that we're wrapping around another high-level compiler.

**Comment by jansel** [comment]
`make format` should fix the lint errors\

**Comment by soumith** [comment]
> @soumith are you ok with the Intel folks exposing a C++-level API from PyTorch core? This will avoid the need to install two copies of MKL because PyTorch hides the symbols. When TorchDynamo is available in core this will become a bit simpler...

This part seems fine if that's what we want.

But more concretely, @jansel what is your long-term vision for a CPU backend for PyTorch? I'd like to see a proper, runtime codegen. But it's not clear to me what a statically compiled C++-level API gives us other than things like hard-coded epilogue fusions.

**Comment by jansel** [comment]
> But more concretely, @jansel what is your long-term vision for a CPU backend for PyTorch? I'd like to see a proper, runtime codegen. But it's not clear to me what a statically compiled C++-level API gives us other than things like hard-coded epilogue fusions.

I agree, the focus should be on more general purpose codegen.

Calling MKL in this way wasn't my idea. I would push stuff like https://github.com/pytorch/torchdynamo/issues/1174 and https://github.com/pytorch/torchdynamo/issues/1175 which I believe the Intel folks are also looking at.

I think having a FX pass above inductor that does pattern-matching style fusions does fit into the larger plan though.  That is how we plan to integrate nvFuser, AIT, cuDNN, etc.  I don't see MKL as very different from those.

**Comment by soumith** [comment]
> I think having a FX pass above inductor that does pattern-matching style fusions does fit into the larger plan though. That is how we plan to integrate nvFuser, AIT, cuDNN, etc. I don't see MKL as very different from those.

I'm pretty supportive of that. I think what threw me off was that this PR isn't integrating above inductor, but within inductor.

**Comment by jgong5** [comment]
> Since we have codepaths to leverage these already (to_mkldnn, etc.), lets focus on long-term for the Inductor backend. It doesn't seem worth it to do the same kind of work again just for short-term inference wins that people can get via other codepaths.
> I'd like to see a proper, runtime codegen. But it's not clear to me what a statically compiled C++-level API gives us other than things like hard-coded epilogue fusions.
> I'm pretty supportive of that. I think what threw me off was that this PR isn't integrating above inductor, but within inductor.

@soumith My two cents. Integrating oneDNN post-ops into the FX pass inside inductor is an intermediate step towards a full-fledged codegen for conv and gemm. There is existing path that leverages oneDNN (a.k.a. mkldnn) but additional fusions with memory-bound operations, like those in BERT (multi-head-attention, add+layernorm), are still missing in the stock PyTorch. We implemented some rigid pattern-matching based TS graph rewrites in IPEX (https://github.com/intel-innersource/frameworks.ai.pytorch.ipex-cpu/blob/cpu-device/csrc/jit/passes/graph_rewrite_mha.cpp#L226) while these things can be more generally supported by inductor. Adding oneDNN post-ops to the inductor path combines the goodness from both oneDNN fusion and inductor's current codegen. In the long run, the C++-level API can expose micro-kernels as the building blocks of conv and gemm. Then, the conv and gemm would be codegen as a loop-nest over the micro-kernels. Does this plan make sense to you?

**Comment by jansel** [comment]
> I'm pretty supportive of that. I think what threw me off was that this PR isn't integrating above inductor, but within inductor.

`fuse_fx()` is effectively "above inductor", since aot_autograd/normalization is called from inductor -- and most of these pattern matchers will to run after that -- inductor will need to also call the pattern matchers.

File: test/inductor/test_torchinductor.py [hunk 6]
```diff
@@ -199,9 +216,15 @@ def upcast_fn(x):
             else:
                 return x
 
+        def get_original_lowp_dtype(example_inputs):
+            dtypes = [x.dtype for x in example_inputs if isinstance(x, torch.Tensor)]
+            dtype_set = set(dtypes)
+            return dtype_set.pop() if len(dtype_set) == 1 else torch.half
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:222]: This is used to support bfloat16 type? What if some of the inputs are bfloat16 (or half) and some are float?

File: test/inductor/test_torchinductor.py [hunk 7]
```diff
@@ -1274,6 +1297,40 @@ def fn(a, b):
             check_lowp=False,
         )
 
+    @unittest.skipIf(HAS_CUDA, "only support bf16 linear unary fusion on cpu")
```
Comment (jgong5) [inline test/inductor/test_torchinductor.py:1300]: If it is CPU only, how about put it inside `if HAS_CPU:` check?

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Comment by jansel** [comment]
We have migrated `torchdynamo` to `torch._dynamo` and will use the `pytorch/pytorch` repo for future development.  Please resubmit this PR to https://github.com/pytorch/pytorch/

More details and instructions to port this PR over can be found in https://github.com/pytorch/torchdynamo/issues/1588

File: torchinductor/ir.py [hunk 8]
```diff
@@ -2943,6 +2943,56 @@ def should_allocate(self):
         return False
 
 
+class LinearEltwise(ExternKernelAlloc):
+    kernel = "torch.ops.mkldnn._linear_pointwise"
```
[the remark this answers is not in the document]
Comment (chunyuan-w) [reply torchinductor/ir.py:2947]: Got it. I'll rename it in the new PR submitted to pytorch repo.
